# PRACTICE 3
[Dataset](https://drive.google.com/open?id=15IyT1ODuDKgZb8WN6iG64hFJRZWJTSAz) : Titanic dataset.

**Latar Belakang** : Dataset berisi data untuk 887 penumpang Titanic. Setiap baris mewakili satu orang. Kolom menggambarkan atribut yang berbeda tentang orang tersebut termasuk apakah mereka selamat (S), usia mereka (A), kelas penumpang mereka (C), jenis kelamin mereka (G) dan tarif yang mereka bayar (X).
<br>

**Tujuan** : Menjawab pertanyaan-pertanyaan yang telah tersedia.

## 0. Load Library

In [17]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

### I. What is the dimension (col, row) of the data frame?

In [20]:
df = pd.read_csv('titanic.csv')

In [21]:
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500

In [22]:
df.shape

(887, 8)

> Dengan menggunakan fungsi *shape* maka dapat diketahui dimensi data. Dapat diektahui terdapat 887 row/baris dan 8 columns/kolom

### II. How to know data type of each variable?

In [8]:
df.dtypes

Survived                     int64
Pclass                       int64
Name                        object
Sex                         object
Age                        float64
Siblings/Spouses Aboard      int64
Parents/Children Aboard      int64
Fare                       float64
dtype: object

> Terdapa 4 feature yang tipe datanya integer, 2 yang float, dan 2 yang string

## III. How many passengers survived (Survived=1) and not-survived (Survived=0)?

In [9]:
df.groupby('Survived').size()

Survived
0    545
1    342
dtype: int64

- Menggunakan fungsi *groupby* amka dapat diketahui 342 penumpang yang selamat (code = 1) dan 545 penumpang yang tidak selamat (code = 0).

### IV. How to drop column ‘Name’ from the data frame?

In [10]:
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500

In [11]:
df.drop(columns=['Name'], inplace=True)

In [23]:
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500

> Menggunakan fungsi df.*drop* maka dapat menghilangkan kolom yang tidak ingin digunakan 

### V. Add one new column called ‘family’ to represent number of family-member aboard (hint: family = sibsp + parch)

In [24]:
df['family'] = df['Siblings/Spouses Aboard'] + df['Parents/Children Aboard']
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male  22.0                        1                        0   7.2500   
1  female  38.0                        1                        0  71.2833   
2  female  26.0                        0                        0   7.9250   
3  female  35.0                        1                        0  53.1000   
4    male  35.0                        0                        0   8.0500   

   family  
0       1  
1       1  
2       0  
3       1  
4       0

> dapat menggunakan simbol '+' untuk menambahan satu kolom baru yang disebut 'family' untuk mewakili jumlah anggota keluarga di atas kapal 

### VI. As shown, columns ‘Age’ contains missing values. Please add new column named ‘Age_miss’ to indicate whether Age is missing or not (Age_miss = ‘YES’ for missing value and ‘NO’ for non-missing value). 


In [25]:
#tidak ada missing values
df.isnull().sum()

Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
family                     0
dtype: int64

In [26]:
age_miss = []
for i in range(len(df)):
    if np.isnan(df['Age'][i]):
        age_miss.append('YES')
    else:
        age_miss.append('NO')

In [27]:
df['age_miss'] = age_miss
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male  22.0                        1                        0   7.2500   
1  female  38.0                        1                        0  71.2833   
2  female  26.0                        0                        0   7.9250   
3  female  35.0                        1                        0  53.1000   
4    male  35.0                        0                        0   8.0500   

   family age_miss  
0       1       NO  
1       1       NO  
2       0       NO  
3       1       NO  
4       0       NO

### VII. Please fill Age missing value with means of existing Age values

In [36]:
df['Age'] = df['Age'].fillna(df['Age'].mean())

> Dengan menggunakan fungsi *fillna* maka dapat mengimputasi variabel missing dengan *.mean* maka dapat mengimputasi dengan rata-rata variabel.

### VIII. What is the maximum passenger Age who survived from the tragedy? 

In [37]:
df[df['Survived']==1]['Age'].max()

80.0

> Dengan menggunakan fungsi *.max* maka dapat mengethaui nilai maksimum dari suatu kolom. Umur maksimum yang dapat selamat dalam tragedi tersebut yaitu 80 tahun.

### IX. How many passengers survived from each ‘PClass’? 

In [38]:
df[df['Survived']==1].groupby('Pclass')['Survived'].count()

Pclass
1    136
2     87
3    119
Name: Survived, dtype: int64

> Dengan menggunakan fungsi *count* maka dapat diektahui jumlah penumpang yang selamat untuk setiap kategori PClass yaitu kategori 1 = 136, kategori 2 = 87 dan kategori 3 = 119.

### X. How to randomly split the data frame into 2 parts (titanic1 and titanic2) with proportion of 0.7 for tttanic1 and 0.3 for titanic2 ? 

In [39]:
# Package untuk Split Data
from sklearn.model_selection import train_test_split

In [40]:
titanic1 = df.sample(frac=0.7).reset_index()
titanic1.head()

index  Survived  Pclass                           Name   Sex   Age  \
0    600         0       3       Mr. Ernst William Torber  male  44.0   
1    756         0       3              Mr. John Garfirth  male  23.0   
2    734         0       3              Mr. Kanio Ivanoff  male  20.0   
3    401         0       3  Mr. Pekka Pietari Hakkarainen  male  28.0   
4    172         0       3       Mr. Antti Wilhelm Sivola  male  21.0   

   Siblings/Spouses Aboard  Parents/Children Aboard     Fare  family age_miss  
0                        0                        0   8.0500       0       NO  
1                        0                        0  14.5000       0       NO  
2                        0                        0   7.8958       0       NO  
3                        1                        0  15.8500       1       NO  
4                        0                        0   7.9250       0       NO

In [41]:
titanic2 = df.drop(titanic1.index)
titanic2.head()

Survived  Pclass                           Name     Sex   Age  \
621         0       3           Mr. David John Bowen    male  21.0   
622         0       1           Mr. Frederick Sutton    male  61.0   
623         0       2  Rev. Charles Leonard Kirkland    male  57.0   
624         1       1   Miss. Gretchen Fiske Longley  female  21.0   
625         0       3       Mr. Guentcho Bostandyeff    male  26.0   

     Siblings/Spouses Aboard  Parents/Children Aboard     Fare  family  \
621                        0                        0  16.1000       0   
622                        0                        0  32.3208       0   
623                        0                        0  12.3500       0   
624                        0                        0  77.9583       0   
625                        0                        0   7.8958       0   

    age_miss  
621       NO  
622       NO  
623       NO  
624       NO  
625       NO

In [42]:
titanic1.shape

(621, 11)

In [43]:
titanic2.shape

(266, 10)

> Cara membagi dataframe secara acak menjadi 2 bagian (titanic1 dan titanic2) dengan proporsi 0,7 untuk titanic1 yaitu dengan menggunakan funsi *df.sample(frac=0.7).reset_index()* dan dikarenakan sudah di *define* titanic1, maka untuk mendapatkan titanic2 dapat melakukan drop titanic1.